### ISSUE TITLE
Multi select with opt groups does not render checkboxes

### ISSUE THREAD

**Issue by christiaanEbit** [Dogfalo/materialize#2522]
If you create a multiple dropdown and add optgroups it does not render the checkboxes.

```
<select name="children[]" multiple id="multi_select_dropdown">
    <option value="" disabled selected>Choose Routes</option>
    <optgroup label="assigned">
        <option value="3" selected>Option 2</option>
        <option value="3" selected>Option 4</option>
    </optgroup>
    <optgroup label="Unassigned">
        <option value="3" selected>Option 3</option>
        <option value="3" selected>Option 5</option>
    </optgroup>
</select>
```

**Comment by TyrionGraphiste** [comment]
For now it's pretty difficult to develop this functionality because index's option are resetted to 0 after each optgroup and we based the generated list item of the ul on these indexes, so with optgroup there is no more correspondence between the elements.

Still, I'm working on a fix but idk if it could be accepted and the fix will take some time.

**Comment by burito1043** [comment]
TyrionGraphiste _ That's awesome you found a fix in your PR. How would I access this?

**Comment by TyrionGraphiste** [comment]
Hi, you could simply overload the material_select component with mine like by declaring a script before the end of your </body> with this content : http://pastebin.com/vsaaB9GD

**Comment by tiagoovieira** [comment]
I'm using your PR code, but when I try to select a second element (I'm using a multiple select) I get an error: `Uncaught TypeError: Cannot read property 'indexOf' of undefined` 

In this line of code:
`index = optionsValuesSelected[indexOptgroup].indexOf(indexOption);`, line 483 in your pastebin

**Comment by TyrionGraphiste** [comment]
Could you provide a JSfiddle or CodeOpen ?

**Comment by burito1043** [comment]
Hello, your script worked, but broke the selects in iOS with the touch
events. I suspect because the file is separate, it's not picking up some of
the script. Any idea when this will be in the public and is there a fix for
this.

Thank you in advance!

-- S

**Comment by TyrionGraphiste** [comment]
Hi,

Thanks for reporting, it's surely because the last changes I did for hide cursor on Safari and IE on several OS. I don't think code separation is the problem.

Well I'll modify my code, sorry for not working.

**Comment by burito1043** [comment]
Awesome! Do you suspect it's a tough fix?

**Comment by TyrionGraphiste** [comment]
For now I can just revert changes from the last push, I'll do that tonight.

**Comment by tiagoovieira** [comment]
@TyrionGraphiste, no problem! Here is the Fiddle, with the materialize.js with your modifications

**Comment by burito1043** [comment]
With that revert, do you suspect that multiple select opt groups will still
work?

**Comment by TyrionGraphiste** [comment]
@tiagoovieira Thanks I'll take a look.
@burito1043 Yes it will, the revert changes only the "fix" for cursor showing on Safari and IE.

**Comment by tiagoovieira** [comment]
@TyrionGraphiste I'm not able to replicate correctly the error on fiddle, although you could find the whole repo [here] (https://github.com/tiagoovieira/reedsy_test) 

But I was digging through your changes and what I understood from it, in the function `returnIndexes` (line 768 on your forms.js PR), when I select the first item `$select.find('option:not(:disabled)')` (line 774) returns the list of all my select options, but when I select a second item, returns an empty list. Can somehow all my options suddenly are disabled?

**Comment by TyrionGraphiste** [comment]
@tiagoovieira Ok no problem. I'm thinking to rewrite the entire component in Vanilla JS because the code has become complicated and a bit maintainable.

**Comment by burito1043** [comment]
Sorry to bug you, but I'm not sure which time zone you're in. Will you let
me know when you revert and how I can get that file again? Thank you in
advance.

**Comment by TyrionGraphiste** [comment]
I reverted since few hours.

**Comment by burito1043** [comment]
I just implemented your reverted version and it didn't change anything. Do
you have any ideas? Thank you.

**Comment by TyrionGraphiste** [comment]
What is your issue exactly ?

**Comment by burito1043** [comment]
Alright, I think I figured it out. Now, my only issue is that when you make
a selection or touch off of the multi select with optgroups, it does not
seen to shift focus from the dropdown and get the values to use. Any ideas
here?

**Comment by burito1043** [comment]
nevermind. false alarm. I'm at a loss here.

**Comment by burito1043** [comment]
Have you made any changes to the optgroup support?

**Comment by TyrionGraphiste** [comment]
Hi, I have almost finished, there is some behaviors unexpected when passing in Vanilla JS, so I'm trying to fix it. I'll do a new pull request once it should be finished.

**Comment by burito1043** [comment]
Wow. Thank you so much.

**Comment by burito1043** [comment]
Sorry to bother you on this, is there an eta on this functionality. I'm
trying to see if the ios bugs will make this viable or if I should consider
another approach.

Thanks in advance.

**Comment by TyrionGraphiste** [comment]
Hi, I'm doing a pastebin for you. Remember this is not the latest version of this functionality.

**Comment by burito1043** [comment]
awesome, thanks. where would I grab that?

**Comment by TyrionGraphiste** [comment]
I edited my post above.

**Comment by burito1043** [comment]
Just checking in...any progress? I'm trying to get this implemented into a
project and have my fingers crossed this will work. I tried your latest
past and still no luck with it working in mobile. Thanks again for all your
work on this.

Respectfully,

**Comment by TyrionGraphiste** [comment]
I (updated)[http://pastebin.com/Ww03xKrT] the component, could you try it and tell me if you get any issue ?
Thank you.

**Comment by burito1043** [comment]
I just tested it and it's yielding an error... TypeError: null is not an
object (evaluating 'parent.querySelector'). I hope this helps.

**Comment by TyrionGraphiste** [comment]
Ok, can you give me the line ?

**Comment by burito1043** [comment]
Sorry. Line 431

**Comment by TyrionGraphiste** [comment]
Ok, I think I've fixed some issues. Could you post your select structure ?

**Comment by burito1043** [comment]
<div class="input-field" id="BeverageWrapper">
    <select id="Beverage" multiple>
        <option value="" disabled selected>Choose your Beverage</option>
    </select>
</div>

<div class="input-field" id="TypesWrapper">
    <select id="Types" multiple>
        <option value="" disabled selected>Choose your Type</option>
    </select>
</div>

<div class="input-field" id="ProductWrapper">
    <select id="Product" multiple="">
        <option value="" disabled selected>Choose your Product</option>
    </select>
</div>

**Comment by burito1043** [comment]
shall I test anything?

**Comment by TyrionGraphiste** [comment]
Hi, I fixed some issues. I'll post in five minutes. If you want to test and tell me if you get another incredible error it will be great.

**Comment by burito1043** [comment]
I'd love to test...what's the pastebin?

**Comment by TyrionGraphiste** [comment]
These are the two latest versions of the component. Look above your last post.

**Comment by burito1043** [comment]
I'm looking at it in email...is there a link?

**Comment by burito1043** [comment]
Found it.

**Comment by burito1043** [comment]
I think this is closer...but my options are not populating now.

**Comment by burito1043** [comment]
Okay, after further testing... 

1. It's not appending my options from JS like before. They are in the DOM, but not in the select menu.
2. The selection of the field seems to be working better.
3. When I try an opt group multi select just in the markup it seems to work.

Hope this helps. I think we're closer.

**Comment by TyrionGraphiste** [comment]
I saw in your select that you don't have options but one disabled. Do you fill your select on load with javascript ?

I think you don't have a normal behavior using the material_select component because all cases are working fine with me.

**Comment by burito1043** [comment]
Yes, I'm filling in the selects with JS on load. Any thoughts? I'm getting desperate. I know this kind of worked, if this sheds any light. 


$('#Format').material_select('destroy')
$('#Format').material_select()

**Comment by TyrionGraphiste** [comment]
Please do a JSfiddle.

**Comment by burito1043** [comment]
I can't do a fiddle since the service we're using to provide the data to
populate the selects is not publicly accessible. With some of the filtering
we're doing, rebuilds the selects and the destroying of them is
conflicting, since they're not real selects, but a manipulation of a DOM
element. At this point, I'm starting to use hope of using these selects,
which is a bummer because I love the framework and the way these feel. Is
there any way to troubleshoot this?

I could send you the js creating the options and optgroups?

**Comment by TyrionGraphiste** [comment]
I just update my code for avoid destroy select they don't need.
I'll do my PR in few time.

**Comment by burito1043** [comment]
great, thanks. any other suggestions with the js not populating the selects?

**Comment by burito1043** [comment]
It's working better... we're getting an error on one of the multiples with
optgroups.

TypeError: undefined is not a function (evaluating
'parent.querySelector('input[type="checkbox"]')')
Line 451.

Any ideas on this?

**Comment by TyrionGraphiste** [comment]
What is your select structure once it's completely filled ? With all options and optgroups.

**Comment by burito1043** [comment]
Here it is once it's populated in the DOM

**Comment by burito1043** [comment]
The first two selects are working great. It's the product one that is
failing, which is the one with optgroups.

**Comment by TyrionGraphiste** [comment]
You didn't write the structure.

**Comment by burito1043** [comment]
Yes, we did. Here is the JS that makes the structure? You can see the
selectors accordingly. Since they are being created dynamically from js,
which gets the data  from a service, I gave you how it resolves in the DOM
to see our structure. Make sense?

**Comment by burito1043** [comment]
Any ideas on this? I've verified that in my main HTML document, I can add
in a multiple select with optgroups with JS or non dynamic with your
example from the materialize css forms page and it yields the same result.
The error is the same...

TypeError: undefined is not a function (evaluating
'parent.querySelector('input[type="checkbox"]')')
Line 451

Thanks.

**Comment by TyrionGraphiste** [comment]
Try with this version, it should be fixed : http://pastebin.com/8TJptq2p

**Comment by burito1043** [comment]
I'm still testing, but initially the error resolves and all of the selects
are now showing. It seems the placeholder disabled value is staying in the
input. I'll keep testing and let you know if it's 100%.

I really appreciate your help with this. I think we're getting close.

**Comment by burito1043** [comment]
Okay did some testing.

Our other developer says, "It seems like nothing more than a couple
variations of setting selected state on options. Once with the property
(newOption.select = true), once by setting the attribute.

If there are no selected states showing and the count of selected items is
not changing as selections are made, then there's not much we can do."

Does this help?

**Comment by burito1043** [comment]
This is also another note he added...

What the select is doing is emptying then repopulating with current values.
What we need is a reliable way to cause the material select to be rebuilt,
preserving optgroup tags, their child nodes, and the selected state of
those child option nodes. That's what we've needed since day one. An
acceptable alternative would be an update method that reliably rebuilds the
material select from current source-select structure and values provided
that it doesn't screw up (alter) those states/values.

I hope this helps

**Comment by burito1043** [comment]
One last note I'm seeing is that it's adding values but if I deselect one,
it doesn't go away.

Thanks.

**Comment by TyrionGraphiste** [comment]
I don't have 'newOption.select = true' in my code, so you shouldn't have the latest version.

There is still an event for update the select, you have to trigger the event like this :
> $('select').val([0, 1]).trigger('update');

And finally there is no more bugs on my side except one with CSS. Maybe you're trying the code under IE , so this is normal that you can't get the normal behavior.

Select and deselect an option is correctly updating the value of the select. Be sure to verify the correct select.

**Comment by burito1043** [comment]
Thanks for the info... I'll check it out.

**Comment by burito1043** [comment]
$('select').val([0, 1]).trigger('update'); ----- this helped some. Now it
replaces the placeholder values withe the correct ones. Is there an API or
any other calls we might be able to make to get it to keep the selected
states when it rebuilds?

**Comment by TyrionGraphiste** [comment]
There is no other calls except the **init**, the **destroy** and the **update**.
The **update** is one I created because it seems to be useful for not destroying and re-initializing the select every time we wanted to set other values.

**Comment by burito1043** [comment]
Thank you. I'm evaluating the changes and will let you know if I have any
other questions. I appreciate the work you've done to get us to this point.

**Comment by TyrionGraphiste** [comment]
Do you have any issue with the component ? Which browser are u using ?

**Comment by burito1043** [comment]
I am still having issues. Despite the ui working, it's as if it's not
actually selecting it. If I add the class to browser default, and select
the same item in the select, it will actually select it. I'm testing in
chrome and safari and it's behaving consistently.

I'm sorry I wish I could explain it better. It's as if the UI around the
LI's is working, but the underlying selects are not.

Does this make sense?

**Comment by TyrionGraphiste** [comment]
So, there are the changes you made on your side which affect the behavior of the component, I can't do more for you given all is working for me.

**Comment by burito1043** [comment]
It's just strange that it worked before the last release. It's as if they were behaving as normal selects and not they are not. Like I said, it work on browser default mode, which leads me to believe there is still a problem. But I'll keep inspecting and debugging. Thanks again.

**Comment by TyrionGraphiste** [comment]
Vanilla JS doesn't implement polyfill to support lot of browsers.
I can't create locals polyfill given this is only a component, it would be very heavy and not optimized.
All polyfill should be global et accessible for each component they need.

So I think the problem is you're using old browser that doesn't support some methods.
The script is working even with IE 10.

**Comment by burito1043** [comment]
I'm working in webkit browsers that are all very recent, namely chrome and
safari.

**Comment by TyrionGraphiste** [comment]
So, if you could provide a JSfiddle for seeing a broken example, it would be nice.

**Comment by burito1043** [comment]
Again, fiddles are hard, since the data source that populates the selects
is not publicly accessible yet. It's literally working perfect, expect it's
like it's not setting the selected state on the backend. It worked with
your previous version just fine and if I do browser default it works fine.
This leads me to believe it's something in the latest version.

**Comment by burito1043** [comment]
So I still cant get it to actually use a value and the focus won't close
the dropdown on modal. I'm not sure what to do at this point. I appreciate
all your work, but feel I'm running out of options to make this component
work.

**Comment by TyrionGraphiste** [comment]
At this point you should implement your own system or get another alternative.

**Comment by burito1043** [comment]
Hey there... Is there an easy way to extend the width of the sidenav menu
on mobile? If I extend it with the documentation, it breaks the black
overlay or the closed state is positioned on canvas. Thank you.

**Comment by Writtscher** [comment]
Sorry, but this is still not working. Select with opt groups does not add a any checkbox. Code is the same as the code from the docs. Version is: 0.97.5

**Comment by TyrionGraphiste** [comment]
Hi,

They didn't accept the PR for whatever reasons, so there is not yet support for multiple select and optgroups.
If you need this -not official- component you can get it directly from my branch at this address : https://github.com/TyrionGraphiste/materialize/blob/master/js/forms.js

**Comment by ryanaddams** [comment]
Any update on this getting merged/fixed in the main branch?

**Comment by Farki** [comment]
Hi guys,

fix is here! Just replace forms.js with this:
https://gist.github.com/Farki/7f4ed52d95ef8f4fd8bf8e3377bb69e2

Credits: IT Ipsos.cz

**Comment by hatumai** [comment]
Any reason this isn't fixed in the main materialize project?  Anyone have any idea what PR there was to fix this?

**Comment by acburst** [comment]
Fixed in 400aca6
